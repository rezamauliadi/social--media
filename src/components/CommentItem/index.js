import React, { Component } from "react";

import { Button, Comment, Label } from "semantic-ui-react";
import UpdateForm from "src/components/UpdateForm";

import API from "src/api";
import getAvatar from "src/helpers/avatar-helper";

class CommentItem extends Component {
  state = {
    showUpdateForm: false,
    deletingComment: false,
    opacity: "1"
  };

  deleteComment = async () => {
    this.setState({ deletingComment: true });

    const { id } = this.props.comment;
    const status = await API.deleteComment(id);

    if (status === 200) {
      this.setState({ opacity: "0", deletingComment: false });
      setTimeout(() => {
        this.props.onDeleteComment(id);
      }, 500);
    }
  };

  afterSubmitForm = updatedComment => {
    this.props.onUpdateComment(updatedComment);
    this.setState({ showUpdateForm: false });
  };

  openUpdateForm = () => {
    this.setState({ showUpdateForm: true });
  };

  closeUpdateForm = () => {
    this.setState({ showUpdateForm: false });
  };

  renderCommentContent = () => {
    const { showUpdateForm } = this.state;
    const { comment } = this.props;

    if (showUpdateForm) {
      return (
        <UpdateForm
          type="comments"
          id={comment.id}
          title={comment.name}
          body={comment.body}
          parentId={comment.postId}
          additionalInfo={comment.email}
          onClose={this.closeUpdateForm}
          afterSubmit={this.afterSubmitForm}
        />
      );
    }

    return (
      <div>
        <div style={{ fontWeight: "bold", fontStyle: "italic" }}>
          {comment.name}
        </div>
        <div>{comment.body}</div>
      </div>
    );
  };

  render() {
    const { opacity, deletingComment, showUpdateForm } = this.state;
    const { comment } = this.props;

    return (
      <div style={{ opacity, transition: "opacity 500ms linear" }}>
        <Comment>
          <Comment.Avatar style={{ height: "35px" }} src={getAvatar(11)} />
          <Comment.Content>
            <Comment.Author as="a">
              <Label size="small">{comment.email.toLowerCase()}</Label>
            </Comment.Author>
            <Comment.Metadata>
              <div>1 day ago</div>
            </Comment.Metadata>
            <Comment.Text>
              {this.renderCommentContent()}
              <div style={{ display: showUpdateForm ? "none" : "block" }}>
                <Button
                  basic
                  size="tiny"
                  icon="pencil"
                  disabled={deletingComment}
                  onClick={() => this.openUpdateForm()}
                />
                <Button
                  basic
                  size="tiny"
                  icon="trash"
                  loading={deletingComment}
                  disabled={deletingComment}
                  onClick={() => this.deleteComment()}
                />
              </div>
            </Comment.Text>
          </Comment.Content>
        </Comment>
      </div>
    );
  }
}

export default CommentItem;
