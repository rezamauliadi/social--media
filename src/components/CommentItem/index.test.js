import React from "react";
import { shallow } from "enzyme";

import CommentItem from "./index.js";
import API from "src/api";

describe("Component: CommentItem", () => {
  const comment = {
    postId: 1,
    id: 1,
    name: "comment name",
    email: "email@email.com",
    body: "comment body"
  };
  const fnMock = jest.fn();
  const props = {
    comment,
    onUpdateComment: fnMock,
    onDeleteComment: fnMock
  };

  it("should match snapshot", () => {
    const component = shallow(<CommentItem {...props} />);
    expect(component).toMatchSnapshot();
  });

  describe("button click", () => {
    it("should call openUpdateForm when update button clicked", () => {
      const component = shallow(<CommentItem {...props} />);
      const instance = component.instance();

      jest.spyOn(instance, "openUpdateForm");
      component
        .find("Button")
        .at(0)
        .simulate("click");
      expect(instance.openUpdateForm).toHaveBeenCalled();
    });

    it("should call deleteComment when delete button clicked", () => {
      const component = shallow(<CommentItem {...props} />);
      const instance = component.instance();

      jest.spyOn(instance, "deleteComment");
      jest.spyOn(instance.props, "onDeleteComment");

      component
        .find("Button")
        .at(1)
        .simulate("click");
      expect(instance.deleteComment).toHaveBeenCalled();
      setTimeout(() => {
        expect(instance.props.onDeleteComment).toHaveBeenCalled();
      }, 500);
    });
  });

  describe("afterSubmitForm", () => {
    it("should call onUpdateComment", () => {
      const component = shallow(<CommentItem {...props} />);
      const instance = component.instance();

      jest.spyOn(instance.props, "onUpdateComment");
      instance.afterSubmitForm();

      expect(instance.props.onUpdateComment).toHaveBeenCalled();
      expect(component.state("showUpdateForm")).toBe(false);
    });
  });

  describe("deleteComment", () => {
    it("should call deleteComment API", done => {
      const component = shallow(<CommentItem {...props} />);
      const instance = component.instance();

      jest.spyOn(API, "deleteComment").mockImplementation(() => 200);
      jest.spyOn(instance.props, "onDeleteComment");
      jest.spyOn(global, "setTimeout");

      instance.deleteComment();
      process.nextTick(() => {
        expect(component.state("opacity")).toBe("0");
        expect(component.state("deletingComment")).toBe(false);

        expect(global.setTimeout).toHaveBeenCalled();
        expect(instance.props.onDeleteComment).toHaveBeenCalled();

        API.deleteComment.mockClear();
        done();
      });
    });
  });

  describe("openUpdateForm", () => {
    it("should set showUpdateForm to true", () => {
      const component = shallow(<CommentItem {...props} />);
      const instance = component.instance();

      instance.openUpdateForm();
      expect(component.state("showUpdateForm")).toBe(true);
    });
  });

  describe("closeUpdateForm", () => {
    it("should set showUpdateForm to false", () => {
      const component = shallow(<CommentItem {...props} />);
      const instance = component.instance();

      instance.closeUpdateForm();
      expect(component.state("showUpdateForm")).toBe(false);
    });
  });
});
