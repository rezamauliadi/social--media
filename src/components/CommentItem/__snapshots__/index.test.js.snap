// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CommentItem should match snapshot 1`] = `
<div
  style={
    Object {
      "opacity": "1",
      "transition": "opacity 500ms linear",
    }
  }
>
  <Comment>
    <CommentAvatar
      src="/images/stevie.jpg"
      style={
        Object {
          "height": "35px",
        }
      }
    />
    <CommentContent>
      <CommentAuthor
        as="a"
      >
        <Label
          size="small"
        >
          email@email.com
        </Label>
      </CommentAuthor>
      <CommentMetadata>
        <div>
          1 day ago
        </div>
      </CommentMetadata>
      <CommentText>
        <div>
          <div
            style={
              Object {
                "fontStyle": "italic",
                "fontWeight": "bold",
              }
            }
          >
            comment name
          </div>
          <div>
            comment body
          </div>
        </div>
        <div
          style={
            Object {
              "display": "block",
            }
          }
        >
          <Button
            as="button"
            basic={true}
            disabled={false}
            icon="pencil"
            onClick={[Function]}
            size="tiny"
          />
          <Button
            as="button"
            basic={true}
            disabled={false}
            icon="trash"
            loading={false}
            onClick={[Function]}
            size="tiny"
          />
        </div>
      </CommentText>
    </CommentContent>
  </Comment>
</div>
`;
